sw/device/lib/dif/dif_csrng.h

```diff
@@ -0,0 +1,543 @@
+   * to zeros if the provided value is less that 12.
+   */
+  size_t seed_material_len;
+  /**
+   * Seed material used in CSRNG
+   */
+  uint32_t seed_material[12];
+} dif_csrng_seed_material_t;
+
+/**
+ * Generated output state.
+ */
+typedef struct dif_csrng_output_status {
+  /**
+   * Set to true when there is cryptographic entropy data available to
+   * read using `dif_csrng_read_output()`.
+   */
+  bool valid_data;
+  /**
+   * Set to true when the cryptographic entropy data available to read
+   * is FIPS/CC compliant at the hardware level.
+   */
+  bool fips_mode;
+} dif_csrng_output_status_t;
+
+/**
+ * A CSRNG interrupt request type.
+ */
+typedef enum dif_csrng_irq {
+  /**
+   * Indicates command request completed.
+   */
+  kDifCsrngIrqCmdReqDone,
+  /**
+   * Indicates that a request for entropy has been made to the entropy
+   * source.
+   */
+  kDifCsrngIrqEntropyReq,
+  /**
+   * Indicates command exception.
+   */
+  kDifCsrngIrqCmdException,
+  /**
+   * Idicates that an internal FIFO or other fatal error occurred.
+   */
+  kDifCsrngIrqFatalError,
+} dif_csrng_irq_t;
+
+/**
+ * A snapshot of the enablement state of the interrupts for CSRNG.
+ *
+ * This is an opaque type, to be used with the `dif_csrng_irq_disable_all()` and
+ * `dif_csrng_irq_restore_all()` functions.
+ */
+typedef uint32_t dif_csrng_irq_snapshot_t;
+
+/**
+ * Creates a new handle for CSRNG.
+ *
+ * This function does not actuate the hardware.
+ *
+ * @param params Hardware instantiation parameters.
+ * @param[out] csrng Out param for the initialized handle.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_csrng_result_t dif_csrng_init(dif_csrng_params_t params,
+                                  dif_csrng_t *csrng);
+
+/**
+ * Configures CSRNG with runtime information.
+ *
+ * This function should need to be called once for the lifetime of `csrng`.
+ *
+ * @param csrng A CSRNG handle.
+ * @param config Runtime configuration parameters.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_csrng_result_t dif_csrng_configure(const dif_csrng_t *csrng,
+                                       dif_csrng_config_t config);
+
+/**
+ * Initializes CSRNG instance with a new seed value.
+ *
+ * `seed_material` is used as specified in NIST SP 800-90Ar1 section
+ * 10.2.1.3.1. See `dif_csrng_entropy_src` for details on how this value
+ * is mixed with the CSRNG seed.
+ *
+ * `seed_material` can be NULL, in which case CSRNG will use a zero
+ * vector instead.
+ *
+ * @param csrng A CSRNG handle.
+ * @param entropy_src_enable Entropy source input enable.
+ * @param seed_material Seed initialization parameters.
+ * @return The result of the operation.
+ */
+DIF_WARN_UNUSED_RESULT
+dif_csrng_result_t dif_csrng_instantiate(
+    const dif_csrng_t *csrng, const dif_csrng_entropy_src_toggle_t *entropy_src_enable,
+    const dif_csrng_seed_material_t *seed_material);
```

Comment:
Please see my comments re. `dif_csrng_entropy_src_toggle_t`. 
```suggestion
dif_csrng_result_t dif_csrng_instantiate(
    const dif_csrng_t *csrng, dif_csrng_toggle_t entropy_src_enable,
    const dif_csrng_seed_material_t *seed_material);
```
Feel free to keep `dif_csrng_entropy_src_toggle_t` if you think it makes the intent clearer but this should be passed by value.